### ISSUE TITLE
Support asynchronity

### ISSUE THREAD

**Issue by khelle** [phpseclib/phpseclib#1068]
Hello everyone. It came into my attention that this library uses heavily methods such as `_send_channel_packet` , `_send_binary_packet`, `_send_channel_packet` and `_get_binary_packet`. These are the methods that execute asynchronous logic in a blocking-manner, wasting tons of potential CPU power. I believe it would be beneficial for this library to implement asynchronity support, and delegate reading and writing sockets and resources to an event loop.

**Comment by terrafrost** [comment]
I plan on redoing large swaths of SSH2 / SFTP at some point. Right now I'm chiefly focused on ASN.1 / DSA / ECDSA support. I'll then refactor (in a BC breaking way) lots of X509.php and after that I'll work on SFTP / SSH2 stuff.

**Comment by raxbg** [comment]
Can we vote for this? I want it as well :)

**Comment by Fauxil-Fox** [comment]
Any update?

**Comment by terrafrost** [comment]
No. I am still working on ECDSA support. I posted a status update recently here:

https://github.com/phpseclib/phpseclib/issues/1082#issuecomment-396122366

Unfortunately, I have not made a lot of progress in the past two weeks because I was on vacation in Alaska.

Anyway, I am willing to adjust my priorities but to do so but I will need financial compensation to the tune of several thousand dollars.

**Comment by cubiclesoft** [comment]
Well I don't know about anybody else, but I whipped up something in a few hours that adds non-blocking support to the interactive shell:

https://github.com/cubiclesoft/php-ssh/blob/master/support/phpseclib/Net/AsyncSSH2.php

Example usage:

https://github.com/cubiclesoft/php-ssh/blob/master/async_test.php

That demonstrates passing window resize packets and running a process at the same time with no ill effects for doing so.

The only problem left is getting PHP hooked up to actual interactivity but the above changes allow for a few possibilities that didn't exist before.

By the way, it is always bad practice to call `stream_select()` on a blocking socket.  Doing so results in very hard to diagnose bugs - for example, the underlying `select()` call can report a readable socket that ends up blocking even when only attempting to read 1 byte of data.  It's a known issue that mostly affects the Linux TCP/IP stack but any network layer like TLS will present the same problem.  IMO, the underlying system `select()` call should return an error and refuse to work when passed a blocking descriptor.  That change alone would fix so many problems with various misunderstandings of socket programming but it would also break a lot of incorrect code.

**Comment by terrafrost** [comment]
First, it'd help if you incorporated your changes into SSH2.php instead of extending. Like it looks like your `_get_channel_packet` method is 95% copied from phpseclib's SSH2.php. What are the changes you made? I can't do `git diff` (or even use GitHub's web interface) to see the changes you made. I could copy / paste your `_get_channel_packet` and phpseclib's SSH2.php's `_get_channel_packet` and put them into their own files and then do a diff on the two files. I could do that for every function you copy / pasted. But that's not as convenient for me.

Second, barring a multi thousand dollar donation I fairly hostile to the idea of doing this any time soon. I'm trying to revamp the crypto primitives for phpseclib right now. ECDSA is done (altho it needs to be integrated into SSH2.php). AEAD ciphers have been added (AES-GCM and ChaCha20-Poly1305). Next step is ECDH. After that I plan on doing a release of 3.0.

In 4.0 I want to completely redo how SSH2.php (and X509.php) work but that's gonna have to be tabled for 4.0. For example, I want to be able to do stuff like this:

https://github.com/phpseclib/phpseclib/issues/479#issuecomment-61917716
https://github.com/phpseclib/phpseclib/pull/655#issuecomment-92600348
https://github.com/phpseclib/phpseclib/issues/1071#issuecomment-391596953

**Comment by terrafrost** [comment]
Looking at your project I get why you're extended it as opposed to modifying the files (altho I do think you'd do well to use Composer) but modifying phpseclib directly would make it easier for me to see how this might be utilized in the 4.0 branch. For that matter I might even just merge your changes in.

But yah - based on my current roadmap's this is definitely something that'd go into the 4.0 release.
